### ISSUE TITLE
Checking time increase in 8.11 and master

### ISSUE THREAD

**Issue by palmskog** [coq-community/atbr#23]
The checking time of the project increased from around 7 minutes in Coq 8.10 to around 18 minutes in 8.11+beta1 and master (as measured by Travis). This could be due to changes in `omega` and deserves further investigation.

**Comment by ppedrot** [comment]
I am on it. I have observed the following files with major time explosions:
- `Model_Languages`: 4.27s →51.09s
- `StrictStarForm`: 9.27s →200.32s
- `DKA_Epsilon`: 15.18s →408.61s
- `DKA_DFA_Language`: 12.28s →309.85s
- `DKA_DFA_Equiv`: 10.67s →198.16s

The rest of the files seem to be in the same ballpark in both versions.

**Comment by ppedrot** [comment]
`Model_Languages` chokes on a rewrite, so it's probably unrelated to the changes in `omega`...

**Comment by ppedrot** [comment]
All the issues seem related to a big slowdown in typeclass resolution in setoid_rewrite.

**Comment by palmskog** [comment]
@ppedrot many thanks for investigating! Maybe it's then hitting the same issue as the recently reported [Bedrock2 regression](https://github.com/coq/coq/issues/11250)?

**Comment by gares** [comment]
Cc @mattam82

**Comment by ppedrot** [comment]
And the winner is...
```
commit a37b6f81a3d3922dc89a179b50494d0bbd7afbf6
Author: Erik Martin-Dorel <erik.martin-dorel@irit.fr>
Date:   Fri Nov 1 00:49:55 2019 +0100

    [ssr] Refactor/Extend of under to support more relations
    
    (namely, [RewriteRelation]s beyond Equivalence ones)
    
    Thanks to @CohenCyril for suggesting this enhancement

 .../06-ssreflect/10022-ssr-under-setoid.rst        |  37 +++++---
 doc/stdlib/hidden-files                            |   1 +
 plugins/ssr/ssrclasses.v                           |   3 +
 plugins/ssr/ssreflect.v                            |  68 ++------------
 plugins/ssr/ssrsetoid.v                            | 103 +++++++++++++++++++--
 plugins/ssr/ssrunder.v                             |  75 +++++++++++++++
 test-suite/ssr/under.v                             |  69 ++++++++++++++
 7 files changed, 275 insertions(+), 81 deletions(-)
 create mode 100644 plugins/ssr/ssrunder.v
```

**Comment by palmskog** [comment]
Hmm, given that the `ssr` plugin is never loaded in ATBR (whose code predates it), how can that commit cause the slowdowns? Am I missing something?

**Comment by ppedrot** [comment]
Hints and in particular typeclass instances are require-global, so it is enough to have somewhere in the dependency path a required file that loads ssr to get those instances. The ssr files do appear in the list of loaded dependencies of ATBR.

This is probably the best example in favour of switching to the import-scoped semantics for hints. It is just a matter of changing one line in the ML source code, but it will be probably worse than the Export-change-of-semantics switch w.r.t. porting Coq developments.

**Comment by ppedrot** [comment]
The culprit is most likely the following line
```
theories/Setoids/Setoid.v:Require Coq.ssr.ssrsetoid.
```

**Comment by gares** [comment]
Cc @erikmd @CohenCyril . The bug is clearly in the way coq loads hints, but here we need at least a workaround.

Any idea why CS search becomes that slower?

**Comment by CohenCyril** [comment]
Blind guess, does 
```coq
Typeclasses Opaque Under_rel.Under_rel.
```
solve the issue?

**Comment by erikmd** [comment]
@CohenCyril it is not applicable (we get `Error: Cannot coerce Under_rel to an evaluable reference`) as `Under_rel.Under_rel` is already opaque (namely, the module `Under_rel` is opaque).

Given we had already noticed in the corresponding PR https://github.com/coq/coq/pull/10022 that the added TCs required some workaround (cf. the `Hint Cut`s), this makes me think [`ssrsetoid.v`](https://github.com/coq/coq/blob/7d1138657904e5fe8ce1899daa001972ba820545/plugins/ssr/ssrsetoid.v) would need another refactoring…

Could @mattam82 or another TC expert take a look? 

FYI the feature at stake depending on the TCs added in this PR 10022 is tested [in this `test-suite` file](https://github.com/coq/coq/blob/7d1138657904e5fe8ce1899daa001972ba820545/test-suite/ssr/under.v#L371-L374).

**Comment by erikmd** [comment]
FYI I've just pushed a commit that significantly simplifies `ssrsetoid.v` (only keeping one TC instance that is strictly necessary for `under`, and cannot trigger TC looping), more details in PR coq/coq#11343

**Comment by palmskog** [comment]
As can be seen in [this job](https://travis-ci.com/coq-community/atbr/jobs/272279289) the build time for ATBR on Coq master is back to normal (5-7 minutes) after the merge of coq/coq#11343. However, I am keeping this issue open until the fix is backported and double-checked for 8.11 as well. Thanks @erikmd @ppedrot @gares!
